### DISCUSSION TITLE
What is the right way of adding same discriminators on multiple modules?

### DISCUSSION THREAD

**Discussion by ilterugur** [GrapeoffJS/kindagoose#147] [Q&A] [answered]
```
KindagooseModule.forFeature([
    { schema: Event, discriminators: [ClickedEvent] },
])
```
I'm adding discriminators to modules like this.
But what should i do if I want to use Event & ClickEvent on another module too? Should i add them twice?

I get this error when I try to add discriminators also on another module:
```
ERROR [ExceptionHandler] Discriminator with name "ClickEvent" already exists
Error: Discriminator with name "ClickEvent" already exists
```

**Comment by findingtheconcept** [comment, marked as answer]
Hello! Thanks to perfect GitHub notifications, I did not see your message, I am sorry. 

All TypeGoose models have only one instance in Nest application. Kindagoose register them in this way:
1. Kindagoose finds all `.forFeature` calls.
2. Then it gets the schema that passed in `.forFeature` arguments.
3. Looks if it already registered in application.

If Kindagoose finds the provider with the same token, it returns it to the module providers, if not, it creates a new instance of model (see [model-factory.ts](https://github.com/GrapeoffJS/kindagoose/blob/master/src/utils/model-factory.ts) and [discriminator-factory.ts](https://github.com/GrapeoffJS/kindagoose/blob/master/src/utils/discriminator-factory.ts)).

Since discriminators are models too, the same rules handle them.

Let's sum up: 
1. Kindagoose uses ONLY ONE instance of Typegoose Model, even if you use `.forFeature` for the same model for several times. I can't say exactly for sure, but most likely, that your application will use the model from the very first call of `.forFeature`, starting from `AppModule`.
2. Discriminators are processed by the same logic.
3. Discriminators are registered separately from models, so even if you register a discriminator deeper than the model was registered after that, they'll also be applied. For instance, if you have registered a model in `AppModule`, and then you registered a discriminator in `AppModule` dependencies, it will be also registered.

So, you don't need to add discriminators twice in every module. The application will start only when all the dependencies are resolved, then, if you have registered a discriminator in one part of the application, it will be accessible from all the other parts after the app is started.

Maybe the logic I just described is counterintuitively, so I think it makes sense to register all the discriminators in `AppModule`, instead of register them only at the moment when you need them.

Hope it was helpful!
